src/daft-local-execution/src/sinks/outer_hash_join_probe.rs

```diff
@@ -377,7 +380,8 @@ impl StreamingSink for OuterHashJoinProbeSink {
                 Ok(StreamingSinkOutput::NeedMoreInput(None))
             }
             _ => {
-                let state = state
+                let mut guard = state.inner.lock().unwrap();
```

Comment:
we can prob put this as a method on `StreamingSinkState` that does the following:
* as_any
* downcasts
* error message
like `fn get_state_mut<T>(&self) -> &mut T`

Reply:
Opted for 
```
pub(crate) fn with_state_mut<T: DynStreamingSinkState + 'static, F, R>(&self, f: F) -> R
    where
        F: FnOnce(&mut T) -> R,
```
since returning `&mut T` will outlive the lifetime of the mutex guard.